src/IO/UseSSL.cpp

```diff
@@ -2,23 +2,61 @@
 
 #include "config.h"
 
+#include <Common/Exception.h>
+
 #if USE_SSL
-#    include <Poco/Net/SSLManager.h>
+#    include <openssl/provider.h>
+#    include <openssl/crypto.h>
 #endif
 
 namespace DB
 {
+
+namespace ErrorCodes
+{
+    extern const int OPENSSL_ERROR;
+}
+
+#if USE_SSL
+OSSL_PROVIDER * DB::UseSSL::default_provider = nullptr;
+OSSL_PROVIDER * DB::UseSSL::legacy_provider = nullptr;
+#endif
+
 UseSSL::UseSSL()
 {
 #if USE_SSL
-    Poco::Net::initializeSSL();
+    if (ref_count++ == 0)
+    {
+        OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, nullptr);
+
+        legacy_provider = OSSL_PROVIDER_load(nullptr, "legacy");
+        if (!legacy_provider)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "Failed to load 'legacy' provider.");
+
+        default_provider = OSSL_PROVIDER_load(nullptr, "default");
+        if (!default_provider)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "Failed to load 'default' provider.");
+    }
 #endif
 }
 
 UseSSL::~UseSSL()
 {
 #if USE_SSL
-    Poco::Net::uninitializeSSL();
+    if (--ref_count == 0)
+    {
+        if (legacy_provider)
```

Comment:
Call me paranoid, but could we please unload the providers in reverse order in which we loaded them?